src/Config/Converters/DataSourceConverterFactory.cs

```diff
@@ -0,0 +1,123 @@
+    }
+
+    private class DataSourceConverter : JsonConverter<DataSource>
+    {
+        // Determines whether to replace environment variable with its
+        // value or not while deserializing.
+        private bool _replaceEnvVar;
+
+        /// <param name="replaceEnvVar">Whether to replace environment variable with its
+        /// value or not while deserializing.</param>
+        public DataSourceConverter(bool replaceEnvVar)
+        {
+            _replaceEnvVar = replaceEnvVar;
+        }
+
+        public override DataSource? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+        {
+            DataSource dataSource = new(DatabaseType.MSSQL, string.Empty, null);
+            if (reader.TokenType is JsonTokenType.StartObject)
+            {
+
+                while (reader.Read())
+                {
+                    if (reader.TokenType is JsonTokenType.EndObject)
+                    {
+                        break;
+                    }
+
+                    if (reader.TokenType is JsonTokenType.PropertyName)
+                    {
+                        string propertyName = reader.GetString() ?? string.Empty;
+                        reader.Read();
+                        switch (propertyName)
+                        {
+                            case "database-type":
+                                dataSource = dataSource with { DatabaseType = EnumExtensions.Deserialize<DatabaseType>(reader.DeserializeString(_replaceEnvVar)!) };
+                                break;
+                            case "connection-string":
+                                dataSource = dataSource with { ConnectionString = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)! };
+                                break;
+                            case "options":
+                                Dictionary<string, JsonElement> optionsDict = new();
+                                while (reader.Read() && reader.TokenType is not JsonTokenType.EndObject)
+                                {
+                                    string key = reader.GetString()!;
+                                    reader.Read();
+                                    JsonElement element;
+                                    if (reader.TokenType is JsonTokenType.String)
+                                    {
+                                        string stringValue = reader.DeserializeString(replaceEnvVar: _replaceEnvVar)!;
+
+                                        using (JsonDocument doc = JsonDocument.Parse($"\"{stringValue}\""))
+                                        {
+                                            element = doc.RootElement.Clone();
+                                        }
+                                    }
+                                    else
+                                    {
+                                        bool boolValue = reader.GetBoolean();
```

Comment:
why are we assuming the element could only be a string or a boolean?

Reply:
anything else would be an error. updated the code to throw exception.